File: csharp/Dson/DsonRepository.cs
#region LICENSE

//  Copyright 2023 wjybxx
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using Wjybxx.Dson.Text;
using Wjybxx.Dson.Types;

namespace Wjybxx.Dson;

/// <summary>
/// 简单的Dson对象仓库实现 -- 提供简单的引用解析功能。
/// </summary>
public class DsonRepository
{
    private readonly Dictionary<string, DsonValue> indexMap = new();
    private readonly List<DsonValue> valueList = new List<DsonValue>();

    public DsonRepository() {
    }

    public DsonArray<string> ToDsonArray() {
        DsonArray<string> dsonArray = new DsonArray<string>(valueList.Count);
        dsonArray.AddAll(valueList);
        return dsonArray;
    }

    public Dictionary<string, DsonValue> IndexMap => indexMap;

    public List<DsonValue> Values => valueList;

    public int Count => valueList.Count;

    public DsonValue this[int index] => valueList[index];

    public DsonRepository Add(DsonValue value) {
        if (!value.DsonType.IsContainerOrHeader()) {
            throw new ArgumentException();
        }
        valueList.Add(value);

        string localId = Dsons.GetLocalId(value);
        if (localId != null) {
            if (indexMap.Remove(localId, out DsonValue? exist)) {
                DsonInternals.RemoveRef(valueList, exist);
            }
            indexMap[localId] = value;
        }
        return this;
    }

    public DsonValue RemoveAt(int idx) {
        DsonValue dsonValue = valueList[idx];
        valueList.RemoveAt(idx); // 居然没返回值...

        string localId = Dsons.GetLocalId(dsonValue);
        if (localId != null) {
            indexMap.Remove(localId, out DsonValue _);
        }
        return dsonValue;
    }

    public bool Remove(DsonValue dsonValue) {
        int idx = DsonInternals.IndexOfRef(valueList, dsonValue);
        if (idx >= 0) {
            RemoveAt(idx);
            return true;
        }
        else {
            return false;
        }
    }

    public DsonValue? RemoveById(string localId) {
        if (localId == null) throw new ArgumentNullException(nameof(localId));
        if (indexMap.Remove(localId, out DsonValue exist)) {
            DsonInternals.RemoveRef(valueList, exist);
        }
        return exist;
    }

    public DsonValue? Find(string localId) {
        if (localId == null) throw new ArgumentNullException(nameof(localId));
        indexMap.TryGetValue(localId, out DsonValue exist);
        return exist;
    }

    public void ResolveReference() {
        foreach (DsonValue dsonValue in valueList) {
            ResolveReference(dsonValue);
        }
    }

    private void ResolveReference(DsonValue dsonValue) {
        if (dsonValue is AbstractDsonObject<string>
            dsonObject) { // 支持header...
            foreach (KeyValuePair<string, DsonValue> entry in dsonObject) {
                DsonValue value = entry.Value;
                if (value.DsonType == DsonType.Reference) {
                    ObjectRef objectRef = value.AsReference();
                    if (indexMap.TryGetValue(objectRef.LocalId, out DsonValue targetObj)) {
                        dsonObject[entry.Key] = targetObj; // 迭代时覆盖值是安全的
                    }
                }
                else if (value.DsonType.IsContainer()) {
                    ResolveReference(value);
                }
            }
        }
        else if (dsonValue is DsonArray<string> dsonArray) {
            for (int i = 0; i < dsonArray.Count; i++) {
                DsonValue value = dsonArray[i];
                if (value.DsonType == DsonType.Reference) {
                    ObjectRef objectRef = value.AsReference();
                    if (indexMap.TryGetValue(objectRef.LocalId, out DsonValue targetObj)) {
                        dsonArray[i] = targetObj;
                    }
                }
                else if (value.DsonType.IsContainer()) {
                    ResolveReference(value);
                }
            }
        }
    }

    //

    public static DsonRepository FromDson(string dsonString, bool resolveRef = false) {
        DsonRepository repository = new DsonRepository();
        using (DsonTextReader reader = new DsonTextReader(DsonTextReaderSettings.Default, dsonString)) {
            DsonValue value;
            while ((value = Dsons.ReadTopDsonValue(reader)) != null) {
                repository.Add(value);
            }
        }
        if (resolveRef) {
            repository.ResolveReference();
        }
        return repository;
    }

    // 解析引用后可能导致循环，因此equals等不实现
    public override string ToString() {
        // 解析引用后可能导致死循环，因此不输出
        return "DsonRepository:" + base.ToString();
    }
}

File: java/src/main/java/cn/wjybxx/dson/DsonRepository.java
package cn.wjybxx.dson;

import cn.wjybxx.dson.internal.DsonInternals;
import cn.wjybxx.dson.types.ObjectRef;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * 简单的Dson对象仓库实现 -- 提供简单的引用解析功能。
 *
 * @author wjybxx
 * date - 2023/6/21
 */
public class DsonRepository {

    private final Map<String, DsonValue> indexMap = new HashMap<>();
    private final DsonArray<String> container;

    public DsonRepository() {
        container = new DsonArray<>();
    }

    public DsonRepository(DsonArray<String> container) {
        this.container = Objects.requireNonNull(container);
        for (DsonValue dsonValue : container) {
            String localId = Dsons.getLocalId(dsonValue);
            if (localId != null) {
                indexMap.put(localId, dsonValue);
            }
        }
    }

    /** 勿修改返回的对象 */
    public Map<String, DsonValue> getIndexMap() {
        return indexMap;
    }

    /** 勿修改返回的对象 */
    public DsonArray<String> getContainer() {
        return container;
    }

    public int size() {
        return container.size();
    }

    public DsonValue get(int idx) {
        return container.get(idx);
    }

    public DsonRepository add(DsonValue value) {
        if (!value.getDsonType().isContainerOrHeader()) {
            throw new IllegalArgumentException();
        }
        container.add(value);

        String localId = Dsons.getLocalId(value);
        if (localId != null) {
            DsonValue exist = indexMap.put(localId, value);
            if (exist != null) {
                DsonInternals.removeRef(container, exist);
            }
        }
        return this;
    }

    public DsonValue removeAt(int idx) {
        DsonValue dsonValue = container.remove(idx);
        String localId = Dsons.getLocalId(dsonValue);
        if (localId != null) {
            indexMap.remove(localId);
        }
        return dsonValue;
    }

    public boolean remove(DsonValue dsonValue) {
        int idx = DsonInternals.indexOfRef(container, dsonValue);
        if (idx >= 0) {
            removeAt(idx);
            return true;
        } else {
            return false;
        }
    }

    public DsonValue removeById(String localId) {
        Objects.requireNonNull(localId);
        DsonValue exist = indexMap.remove(localId);
        if (exist != null) {
            DsonInternals.removeRef(container, exist);
        }
        return exist;
    }

    public DsonValue find(String localId) {
        Objects.requireNonNull(localId);
        return indexMap.get(localId);
    }

    public DsonRepository resolveReference() {
        for (DsonValue dsonValue : container) {
            resolveReference(dsonValue);
        }
        return this;
    }

    private void resolveReference(DsonValue dsonValue) {
        if (dsonValue instanceof AbstractDsonObject<?> dsonObject) { // 支持header...
            for (Map.Entry<?, DsonValue> entry : dsonObject.entrySet()) {
                DsonValue value = entry.getValue();
                if (value.getDsonType() == DsonType.REFERENCE) {
                    ObjectRef objectRef = value.asReference();
                    DsonValue targetObj = indexMap.get(objectRef.getLocalId());
                    if (targetObj != null) {
                        entry.setValue(targetObj);
                    }
                } else if (value.getDsonType().isContainer()) {
                    resolveReference(value);
                }
            }
        } else if (dsonValue instanceof DsonArray<?> dsonArray) {
            for (int i = 0; i < dsonArray.size(); i++) {
                DsonValue value = dsonArray.get(i);
                if (value.getDsonType() == DsonType.REFERENCE) {
                    ObjectRef objectRef = value.asReference();
                    DsonValue targetObj = indexMap.get(objectRef.getLocalId());
                    if (targetObj != null) {
                        dsonArray.set(i, targetObj);
                    }
                } else if (value.getDsonType().isContainer()) {
                    resolveReference(value);
                }
            }
        }
    }

    //
    public static DsonRepository fromDson(DsonReader reader) {
        return fromDson(reader, false);
    }

    public static DsonRepository fromDson(DsonReader reader, boolean resolveRef) {
        try (reader) {
            DsonRepository repository = new DsonRepository(Dsons.readTopContainer(reader));
            if (resolveRef) {
                repository.resolveReference();
            }
            return repository;
        }
    }

    // 解析引用后可能导致循环，因此equals等不实现
    @Override
    public String toString() {
        // 解析引用后可能导致死循环，因此不输出
        return "DsonRepository:" + super.toString();
    }

}